### ISSUE TITLE
feedback on coordination

### ISSUE THREAD

**Issue by kirkshoop** [Reactive-Extensions/RxCpp#9]
The goal of `coordination` is to remove all synchronization from the 'join' and time operators. Re-implementing an optimized serialization of calls to multiple subscribers in each operator obscures the functionality, duplicates the same work over and over again and slows down the cases where all the sources are already serialized (UI events etc..). 

An operator passes all the sources through the optional coordination and uses the returned observables instead. These are expected to never overlap calls across all the observers subscribed to them. 

This results in multiple implementations of `coordination` that accomplish the goal in different ways with different tradeoffs, but the operators do not care. The default impl of `coordination` passes the original source through unchanged which results in the default having no synch overhead.

**Comment by kirkshoop** [comment]
I have started experimenting with removing the coordination, recursion and schedulable. 

Http://github.com/kirkshoop/rxcpp/simplescheduler has the changes. 

This leaves a simplified structure of scheduler, worker and action. These are then used by helpers to coordinate multiple observes using the same worker to ensure serialization. 

Operators like merge use the helpers and have no explicit syncronization.

**Comment by kirkshoop** [comment]
Corrected link

https://github.com/kirkshoop/RxCpp/tree/simplescheduler

**Comment by BenPope** [comment]
I don't have any time right now to comment on this (or really look at it at all), but I would like to mention: https://github.com/chriskohlhoff/executors
Which is an implementation of: http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2014/n4242.html
I think there is pretty strong support for this proposal, so something that's sort of compatible might be a good idea, or at least, it may provide you some ideas in your refactoring.  I'll take a closer look on the weekend. This may not be relevant at all, just wanted to point it out.

**Comment by kirkshoop** [comment]
@BenPope I have contributed to a different [impl](https://github.com/arturl/executors) of a different executor paper. my main issue with the current papers is that they do not allow each executor instance to have a local sense of time (virtual_time scheduling is the basis of all the rxcpp tests)

**Comment by furuholm** [comment]
I think that it makes sense to split the worker into a worker and an executor. This would allow for reusing executors, ease the integration of external executors and to switch executor depending on the context. E.g. an event loop scheduler could either use a "single thread executor" or a "thread pool executor".

**Comment by kirkshoop** [comment]
An interesting idea. 

TLDR; while given the std approach to executors, it is inevitable that workers will use executors, this composition will impact the implementation and performance of workers negatively. As I have said before, the std executors must support virtual-timed execution or building async algorithms using them will be ugly and hard or impossible to test

A worker guarantees a serial execution of the items scheduled on it.

The event loop workers guarantee that a limited number of workers are running items in parallel (each worker's items are still run serially).

The existing event loop does this by creating N new threads and assigning each new worker to a thread in naive round-robin.

If the new event loop used a thread-pool executor it would have to use a semaphore to limit concurrency instead.

If a semaphore based event loop was using a new-thread executor that created a new thread per instance of the executor or per item, it would not have the same perf characteristics as the current implementation since each worker would have its own thread instead of sharing N threads. Many more threads would be created and destroyed even though the same concurrency limits were being enforced.

I would need a round-robin executor to get the old semantic - at which point the semaphore event loop is duplicating effort by limiting to N concurrency twice - especially when (round-robin N < semaphore N).

Does the executor allow timed execution? 

A semaphore event loop that executes one function that loops on a queue and sleeps until the head item in the queue is ready (has a time => now()) will not be appropriate for a thread-pool executor without timed execution. Whether timed execution is available or not, the event loop would need to execute recursively rather than sleep a thread-pool thread until the head of the queue is ready. 

However, if event loop executes recursively on a new-thread or round-robin executor the executors would need additional complexity to be efficient (work-stealing and perhaps priority-queueing) at which point they better be implemented as concurrency limited thread-pools (if implemented from scratch they will just be bad thread-pools). Of course, the semaphore event loop using a thread-pool executor is a concurrency limited thread-pool, so when would any executor other than thread-pool be used with semaphore event-loop?

As I have said before, the std executors must support virtual-timed execution or building async algorithms using them will be ugly and hard or impossible to test.

**Comment by furuholm** [comment]
Reading my comment again I realize that I was quite unclear. I'm actually not suggesting that std executors should be used, instead I just mean that workers can be separated into two classes: worker and executors. One such specialized executor could be a bounded queue exectutor that do not do any allocations on the heap (I'm not sure that is possible, but it seems reasonable). This would be suitable for hard real time systems that can guarantee that the queue will never be filled.

This is btw how at least some of the schedulers in RxJava has been implemented where an executor from the Java standard library is used.

**Comment by kirkshoop** [comment]
closing this, discussion on scheduling design will be an open issue in v3 now.
